Refactor: database actions + fix v4 transition

datawhores opened the pull request:
1. use upsert method instead of catching integrity error, to reduce the number of sqlite write transactions
2. this also fixes the hash directory option in the menu
Note: the sorting option is still disabled currently
3. This commit tries will also try to fix v4 transition

NTFSvolume commented:
Using execute many for the hash iserts would not yield any real performance improvement. The file reading is asynchronus but the computation of the hash itself is synchronus.

There are never 2 hashes that need to be added to the db at the same time.

Reading all the hashes at startup would. It would remove basically every need to read the db after and only leave insertion of new hashes.

The memory footprint will be higher but even with a million unique hashes it would be less than 100MB

datawhores commented:
> Using execute many for the hash iserts would not yield any real performance improvement. The file reading is asynchronus but the computation of the hash itself is synchronus.
> 
> There are never 2 hashes that need to be added to the db at the same time.
> 
> Reading all the hashes at startup would. It would remove basically every need to read the db after and only leave insertion of new hashes.
> 
> The memory footprint will be higher but even with a million unique hashes it would be less than 100MB


I modified the hash_table class to update the database with all the generated information together
I made a branch to test if there was any speed difference in the two methods

```
def generate_random_values():
    """
    Generates random values for the given function:
        manager.db_manager.hash_table.queue_hash_db(hash_value, "xxhash", file, original_filename, referer)

    Returns:
        A tuple containing the generated random values:
        (hash_value, file, original_filename, referer)
    """

    # Generate random hash_value (example: hexadecimal string)
    hash_value = "".join(random.choices(string.hexdigits, k=32))

    # Generate random file path (example: using random words)
    with tempfile.NamedTemporaryFile(delete=False) as _:
        file = Path(_.name)
    # Generate random original_filename (similar to file)
    original_filename = "test"

    # Generate random referer URL (example: simplified URL)
    referers = ["google.com", "example.com", "wikipedia.org", "youtube.com"]
    referer = f"https://{random.choice(referers)}/"

    return hash_value, file, original_filename, referer
async def batch_helper(manager, num):
    await manager.async_db_hash_startup()
    for _ in range(num):
        hash_value, file, original_filename, referer = generate_random_values()
        await manager.db_manager.hash_table.queue_hash_db(hash_value, "xxhash", file, original_filename, referer)
        os.remove(file)
    await manager.db_manager.hash_table.batch_insert_or_update_hash_db()
    await manager.async_db_close()


async def iterative_helper(manager, num):
    await manager.async_db_hash_startup()
    for _ in range(num):
        hash_value, file, original_filename, referer = generate_random_values()
        await manager.db_manager.hash_table.insert_or_update_hash_db(
            hash_value, "xxhash", file, original_filename, referer
        )
        os.remove(file)
    await manager.async_db_close()


def hash_directory_scanner(manager: Manager, path: Path, number: int = 10):
    """
    Times the execution of the hash_directory_scanner function using timeit.

    Args:
        manager: The manager object.
        path: The path to the directory to hash.
        number: The number of times to run the function (default: 10).
    """

    def wrapper(num):
        asyncio.run(batch_helper(manager, num))

    def wrapper2(num):
        asyncio.run(iterative_helper(manager, num))

    for num in [100, 300, 500, 1000, 5000]:
        result = timeit.timeit(partial(wrapper2, num), number=number)
        print(f"Average execution time for iterative hash_directory_scanner @{num}: {result / number:.4f} seconds")
        result = timeit.timeit(partial(wrapper, num), number=number)
        print(f"Average execution time for batch hash_directory_scanner@ {num}: {result / number:.4f} seconds")
        pass
    pass

```

NTFSvolume commented:
There are 2 flaws in the test:
1. Both tests are actually running synchronously cause they are inside a single task in a single async loop
2. The test does not take into account the compute time of the hash
 
Bulk insertion is faster but it's not really meaningful, even when running the test asynchronously. The bulk of the wait time is the hashing compute itself, which it orders of magnitude slower that the database insertion.

The insertion time itself does not matter cause the insertion is asynchronous. When CDL tries to insert a hash in the table, it intermediately starts hashing the next file. It does not wait for the insertion to finish.

## 1. Running the insertions asynchronously

I modified the test to run the insertions asynchronously, like CDL does. Just need to use async gather inside `batch_helper` and `iterative_helper`:

```python
async def batch_helper(manager: Manager, num):
    await manager.async_db_hash_startup()
    tasks = []
    for _ in range(num):
        hash_value, file, original_filename, referer = generate_random_values()
        tasks += [
            manager.db_manager.hash_table.queue_hash_db(hash_value, "xxhash", file, original_filename, referer),
        ]
    await asyncio.gather(*tasks)
    await manager.db_manager.hash_table.batch_insert_or_update_hash_db()
    await manager.async_db_close()


async def iterative_helper(manager: Manager, num):
    await manager.async_db_hash_startup()
    tasks = []
    for _ in range(num):
        hash_value, file, original_filename, referer = generate_random_values()
        tasks += [
            manager.db_manager.hash_table.insert_or_update_hash_db(
                hash_value, "xxhash", file, original_filename, referer
            )
        ]
    await asyncio.gather(*tasks)
    await manager.async_db_close()

```

These are the results:
```
Average execution time for iterative hash_directory_scanner @100: 0.2606 seconds
Average execution time for batch hash_directory_scanner@ 100: 0.1192 seconds
Average execution time for iterative hash_directory_scanner @300: 0.5795 seconds
Average execution time for batch hash_directory_scanner@ 300: 0.2277 seconds
Average execution time for iterative hash_directory_scanner @500: 0.9017 seconds
Average execution time for batch hash_directory_scanner@ 500: 0.3158 seconds
Average execution time for iterative hash_directory_scanner @1000: 1.5746 seconds
Average execution time for batch hash_directory_scanner@ 1000: 0.5462 seconds
Average execution time for iterative hash_directory_scanner @5000: 7.4400 seconds
Average execution time for batch hash_directory_scanner@ 5000: 2.3313 seconds
```

Even with 5000 insertions the total gain was just a bit over 5s

## 2. Including the actual hashing of the file

I created a 100MB temp file:

```bash
#!/bin/bash

mkdir -p /mnt/ramdisk
mount -t tmpfs -o size=100M tmpfs /mnt/ramdisk

dd if=/dev/zero of=/mnt/ramdisk/tempfile bs=1M count=100
```

Then modified the test to hash that file before inserting a hash into the database:

```python
fake_file = Path("/mnt/ramdisk/tempfile")


async def hash_and_insert(manager: Manager):
    hash_value, file, original_filename, referer = generate_random_values()
    await manager.hash_manager.hash_file(fake_file, "xxh128")
    await manager.db_manager.hash_table.insert_or_update_hash_db(hash_value, "xxhash", file, original_filename, referer)


async def hash_and_insert_batch(manager: Manager):
    hash_value, file, original_filename, referer = generate_random_values()
    await manager.hash_manager.hash_file(fake_file, "xxh128")
    await manager.db_manager.hash_table.queue_hash_db(hash_value, "xxhash", file, original_filename, referer)


async def batch_helper(manager: Manager, num):
    await manager.async_db_hash_startup()
    tasks = []
    for _ in range(num):
        tasks += [hash_and_insert_batch(manager)]
    await asyncio.gather(*tasks)
    await manager.db_manager.hash_table.batch_insert_or_update_hash_db()
    await manager.async_db_close()


async def iterative_helper(manager: Manager, num):
    await manager.async_db_hash_startup()
    tasks = []
    for _ in range(num):
        tasks += [hash_and_insert(manager)]
    await asyncio.gather(*tasks)
    await manager.async_db_close()
```

These are the results:
```
Average execution time for iterative hash_directory_scanner @100: 6.4001 seconds
Average execution time for batch hash_directory_scanner@ 100: 6.0043 seconds
Average execution time for iterative hash_directory_scanner @300: 16.9755 seconds
Average execution time for batch hash_directory_scanner@ 300: 16.7424 seconds
Average execution time for iterative hash_directory_scanner @500: 28.4553 seconds
Average execution time for batch hash_directory_scanner@ 500: 28.5446 seconds
Average execution time for iterative hash_directory_scanner @1000: 62.4587 seconds
Average execution time for batch hash_directory_scanner@ 1000: 60.6429 seconds
```

In some cases the batch insertion is actually slower cause now we are IO bound while reading the file for hashing, either by the filesystem, the CPU or the RAM. In my case it was the ram cause i don't have enough memory to hash that many files at the same time (even thought it was actually the same file). I had to stop it at 1000.

The key difference here is that including the hashing makes the average time about 30x slower with regular insertions and about 60x to 100x slower when using batches. Hashing larger files will be even slower. Hashing compute time is O(N) but database insertion time is O(1).

This test starts the hashing at the same time for all tasks but in reality hashing will start at different times because files do not finish downloading at the exact same moment.

That's why insertion time does not matter. By the time a new hash it computed, all the pending insertions would have already finished because hashing is way slower.

datawhores commented:
Yeah it doesn't really make much of a difference
```
Reading all the hashes at startup would. It would remove basically every need to read the db after and only leave insertion of new hashes.

The memory footprint will be higher but even with a million unique hashes it would be less than 100MB
```

I think I would leave this for another pull requests

I remove the executemany commits/code
@NTFSvolume

datawhores commented:
This commit should fix 
https://github.com/jbsparrow/CyberDropDownloader/issues/450

datawhores commented:
update

The script should warn users if they haven't gone through the proper method for upgrading from v4

NTFSvolume commented on cyberdrop_dl/clients/hash_client.py:
```diff
@@ -58,36 +48,49 @@ async def startup(self) -> None:
-    async def _hash_item(self, file: Path | str, original_filename: str, referer: URL, hash_type=None) -> str:
-        """Generates hash of a file."""
-        key = self._get_key_from_file(file)
+    async def update_db_and_retrive_hash(self, file: Path | str, original_filename: str, referer: URL):
```
```suggestion
    async def update_db_and_retrive_hash(self, file: Path | str, original_filename: str | None = None, referer: URL | None = None):
```

NTFSvolume commented on cyberdrop_dl/clients/hash_client.py:
```diff
@@ -58,36 +48,49 @@ async def startup(self) -> None:
                 raise NotADirectoryError
             for file in path.rglob("*"):
-                await self._hash_item_helper(file, None, None)
+                await self.update_db_and_retrive_hash(file, None, None)
```
```suggestion
                await self.update_db_and_retrive_hash(file)
```

NTFSvolume commented on cyberdrop_dl/clients/hash_client.py:
```diff
@@ -58,36 +48,49 @@ async def startup(self) -> None:
+        return hash
+
+    async def _update_db_and_retrive_hash_helper(
+        self, file: Path | str, original_filename: str, referer: URL, hash_type=None
```
```suggestion
        self, file: Path | str, original_filename: str | None = None, referer: URL | None = None, hash_type: str | None = None
```

NTFSvolume commented on cyberdrop_dl/clients/hash_client.py:
```diff
@@ -112,25 +115,14 @@ async def _hash_item(self, file: Path | str, original_filename: str, referer: UR
         return hash
 
-    async def hash_item(self, media_item: MediaItem) -> None:
+    def save_hash_data(self, media_item, hash):
```
```suggestion
    def save_hash_data(self, media_item: MediaItem, hash: str) -> None:
```

NTFSvolume commented on cyberdrop_dl/managers/manager.py:
```diff
@@ -208,11 +205,18 @@ def args_logging(self) -> None:
-        await self.db_manager.close()
         if not isinstance(self.client_manager, Field):
             await self.client_manager.close()
+        if not isinstance(self.db_manager, Field):
+            await self.db_manager.close()
         await self.cache_manager.close()
         self.db_manager: DBManager = field(init=False)
         self.cache_manager: CacheManager = field(init=False)
```
```suggestion
        await self.async_db_close()
```

NTFSvolume commented on cyberdrop_dl/utils/transfer/db_setup.py:
```diff
@@ -27,61 +23,15 @@ def transfer_v5_to_new_hashtable(self):
-            transfer_v4_db(
-                constants.APP_STORAGE / "download_history.sqlite",
-                constants.APP_STORAGE / "Cache" / "cyberdrop.db",
+        cache_file = constants.APP_STORAGE / "Cache" / "cache.yaml"
+        if not cache_file.is_file() or not self.check_cache_for_moved(cache_file):
+            raise Exception(
+                "Transitioning from 4.x.x is now longer supported on since 6.3.0: read https://script-ware.gitbook.io/cyberdrop-dl/transition-to-v6#import-v4-config-to-v6"
```
Checking the cache it's now enough. Any kind of cache should be able to be deleted and the program still be able to function.

Let CDL fail with the default warning from the v6 validation error or check the config to see if it contains any of the old v4 option names

NTFSvolume commented on cyberdrop_dl.profiling:
Delete this file

NTFSvolume commented on cyberdrop_dl/managers/manager.py:
```diff
@@ -66,15 +66,10 @@ def startup(self) -> None:
         if not self.parsed_args.cli_only_args.appdata_folder:
-            self.first_time_setup.transfer_v4_to_v5()
-
+            self.first_time_setup.check_transfer_from_v4()
```
This check make more sense inside the config manager startup

NTFSvolume commented on cyberdrop_dl/managers/manager.py:
```diff
@@ -127,13 +122,15 @@ async def async_db_hash_startup(self) -> None:
         if not isinstance(self.db_manager, DBManager):
             self.db_manager = DBManager(self, self.path_manager.history_db)
             await self.db_manager.startup()
+        self.first_time_setup.transfer_v5_to_new_hashtable()
```
Try to convert this to an async function and move it to the db_manager startup.

The first time setup module shouldn't be necessary

NTFSvolume replied:
Actually, the entire first_time_setup module should be deleted

NTFSvolume reviewed (commented):
Also please try to save the queries to a variable before executing them cause they are really long and it's hard to read

NTFSvolume commented on cyberdrop_dl/config_definitions/config_settings.py:
```diff
@@ -188,6 +188,7 @@ class DupeCleanupOptions(BaseModel):
     add_md5_hash: bool = False
     add_sha256_hash: bool = False
     send_deleted_to_trash: bool = True
+    batch_hash_inserts: bool = False
```
This comment is outdated

datawhores commented:
Yeah I'm holding off on this until the v4 transfer is fixed, it doesn't work on the current version

datawhores commented:
@NTFSvolume 
- added feature to switch the default config, when switching the config in the menu. 
currently on the loaded config in memory is changed
- fixed issue with hash_directory scanner crashing the program, there was also a smaller issue with the progress bar not updating
- fix v4 transition issues
   * previously it crashed because config manager is required
   * also my db file was in what platformdirs calls the user_data_dir, so I added this

NTFSvolume commented on .gitignore:
```diff
@@ -7,6 +7,12 @@
+.config.yaml
+/build
+
+#config files
+*.yaml
+*.yml
```
YAML files cant be ignored

NTFSvolume commented on cyberdrop_dl/managers/manager.py:
```diff
@@ -241,16 +241,23 @@ def args_logging(self) -> None:
         if not isinstance(self.client_manager, Field):
             await self.client_manager.close()
-        await self.cache_manager.close()
+        if not isinstance(self.db_manager, Field):
+            await self.db_manager.close()
```
```suggestion
        await self.async_db_close()
```

NTFSvolume commented on cyberdrop_dl/managers/manager.py:
```diff
@@ -241,16 +241,23 @@ def args_logging(self) -> None:
-        self.db_manager: DBManager = field(init=False)
-        self.cache_manager: CacheManager = field(init=False)
         self.hash_manager: HashManager = field(init=False)
+        self.db_manager: DBManager = field(init=False)
```
```suggestion
```

NTFSvolume commented on cyberdrop_dl/ui/program_ui.py:
```diff
@@ -113,9 +113,11 @@ def _import_from_v4(self) -> None:
-        if not new_config:
-            return
-        transfer_v4_config(self.manager, *new_config)
+        new_config_name, old_config_path = new_config
```
```suggestion
        if not new_config:
            return
        new_config_name, old_config_path = new_config
```

NTFSvolume commented on cyberdrop_dl/ui/prompts/user_prompts.py:
```diff
@@ -105,6 +105,25 @@ def select_config(configs: list) -> str:
     )
 
 
+def switch_default_config(config_name) -> str:
```
```suggestion
def switch_default_config_to(config_name) -> str:
```

NTFSvolume commented on cyberdrop_dl/ui/prompts/user_prompts.py:
```diff
@@ -105,6 +105,25 @@ def select_config(configs: list) -> str:
+    )
+
+
+def default_config_bool() -> str:
```
```suggestion
def switch_default_config() -> str:
```

NTFSvolume commented on cyberdrop_dl/ui/prompts/user_prompts.py:
```diff
@@ -105,6 +105,25 @@ def select_config(configs: list) -> str:
+    )
+
+
+def active_config_bool(config) -> str:
```
```suggestion
def activate_config(config) -> str:
```

NTFSvolume commented on cyberdrop_dl/ui/prompts/user_prompts.py:
These functions could have better names

datawhores commented:
@NTFSvolume 
You want to take another look when your ready
